### ISSUE TITLE
Dcs-bios + cheap pots

### ISSUE THREAD

**Issue by Nasty76** [dcs-bios/dcs-bios#72]
Hey again , i have test my system and get some issue with a10c trim pot , i use cheapest pot so that not keep accurate position, example pot value is 3200 but that pot give value range 3190-3210 so dcs bios spam message nonstop which cause my buttons not register sometimes position , that help i adjust little pot again , i have my self little play and code arduino and raspi and my mind come is that hard to rewrite pot logic example when value is bigger than something example 200 or something dcs bios sent command or pots value read only 5-10 times per sec?. ofc simple way is buy better pots which not jump when not turn but cheap is allways cheap and my opion is if pot not spam noise messages.
this issue hit me when i try use arduino direct usb my game comp and raspierry pi system too.
if you make some range which not trigger udp. message maybe that can be adjust some arduino sketch or something so people can adjust that correct depend which kinda pots they have.

**Comment by Nasty76** [comment]
oh i look that arduino libr skecth , there is section:
void Potentiometer::pollInput() {
	unsigned int state = map(analogRead(pin_), 0, 1023, 0, 65535);
	if (state != lastState_) {
		char buf[6];
		utoa(state, buf, 10);
		sendDcsBiosMessage(msg_, buf);
	}
	lastState_ = state;
}
i can make my self this some timer based or value change based but maybe i wait ur opion is that right issue or i have make something wrong which cause pots issue.

**Comment by jboecker** [comment]
Correct, the current potentiometer code is too naive for practical applications. 
It's from a time before the whole copy-and-paste stuff from the control reference docs thing, back when the Arduino library was meant more as a minimal example of how to use the ProtocolParser class instead of "all you need to make this work".

See also dcs-bios/dcs-bios-arduino-library/#12 which is a pull request that adds a deadband. It applies to the current development version, but shouldn't be too hard to apply to v1.4.x.

I really need to take a weekend to tidy up the issue tracker and make a new version of the Arduino library, but I probably won't find the time to do it until November.

**Comment by Nasty76** [comment]
Thanks for info i can now week ens rewrite some idea that pot code and rest if that work better. 

----- Alkuperäinen viesti -----
Lähettäjä: "jboecker" <notifications@github.com>
Lähetetty: ‎11.‎9.‎2015 16:26
Vastaanottaja: "dcs-bios/dcs-bios" <dcs-bios@noreply.github.com>
Kopio: "Nasty76" <nasty76@gmail.com>
Aihe: Re: [dcs-bios] Dcs-bios + cheap pots (#72)

Correct, the current potentiometer code is too naive for practical applications. 
It's from a time before the whole copy-and-paste stuff from the control reference docs thing, back when the Arduino library was meant more as a minimal example of how to use the ProtocolParser class instead of "all you need to make this work".
See also dcs-bios/dcs-bios-arduino-library/#12 which is a pull request that adds a deadband. It applies to the current development version, but shouldn't be too hard to apply to v1.4.x.
I really need to take a weekend to tidy up the issue tracker and make a new version of the Arduino library, but I probably won't find the time to do it until November.

**Comment by Nasty76** [comment]
Hey  again i make little buble cum patent which fix pot flood issue :D. use if u want or laugh freely my amateur arduino code :D 

code is here:
#include "Arduino.h"
#include "Potentiometers.h"
int lastsensorValue = 0;
int sensorValue = 0;
namespace DcsBios {
	
	Potentiometer::Potentiometer(char* msg, char pin) {
		msg_ = msg;
		pin_ = pin;
		pinMode(pin_, INPUT);
		lastState_ = map(analogRead(pin_), 0, 1023, 0, 65535);
	}
	void Potentiometer::pollInput() {
	 sensorValue = analogRead(pin_);
	  if(sensorValue< lastsensorValue-3 || sensorValue> lastsensorValue+3 )
  {
		unsigned int state = map(analogRead(pin_), 0, 1023, 0, 65535);
		if (state != lastState_) {
			char buf[6];
			utoa(state, buf, 10);
			sendDcsBiosMessage(msg_, buf);
		}
		lastState_ = state;
	lastsensorValue = sensorValue;
	}}

}

**Comment by Nasty76** [comment]
better fix where jitter is variable , and commented what that line do:
#include "Arduino.h"
#include "Potentiometers.h"
int lastsensorValue = 0;
int sensorValue = 0;
int jitter = 3; // Adjust this number to match your pot jitter , i use blank arduino sketch to see how much my analogread jump allover place and see that was 1-2 depend where my pot is so i put there 3 which eliminate pot spam to dcs game
namespace DcsBios {
	
	Potentiometer::Potentiometer(char* msg, char pin) {
		msg_ = msg;
		pin_ = pin;
		pinMode(pin_, INPUT);
		lastState_ = map(analogRead(pin_), 0, 1023, 0, 65535);
	}
	void Potentiometer::pollInput() {
	 sensorValue = analogRead(pin_);
	  if(sensorValue< lastsensorValue-jitter || sensorValue> lastsensorValue+jitter )
  {
		unsigned int state = map(analogRead(pin_), 0, 1023, 0, 65535);
		if (state != lastState_) {
			char buf[6];
			utoa(state, buf, 10);
			sendDcsBiosMessage(msg_, buf);
		}
		lastState_ = state;
	lastsensorValue = sensorValue;
	}}

}

**Comment by OlHall** [comment]
This pretty much what I wrote in my pull request months ago.  I've since moved on and am writing my own integration system.

**Comment by Nasty76** [comment]
Oh ok nice, i dont have any glue how github work , and if i need somethink i try solve that asap :).  

----- Alkuperäinen viesti -----
Lähettäjä: "Oliver Hall" <notifications@github.com>
Lähetetty: ‎12.‎9.‎2015 10:13
Vastaanottaja: "dcs-bios/dcs-bios" <dcs-bios@noreply.github.com>
Kopio: "Nasty76" <nasty76@gmail.com>
Aihe: Re: [dcs-bios] Dcs-bios + cheap pots (#72)

This pretty much what I wrote in my pull request months ago. I've since moved on and am writing my own integration system.

**Comment by Nasty76** [comment]
Oh ok nice, i dont have any glue how github work , and if i need somethink i try solve that asap :).  I test more that dcs bios and see if there is some more issues. I have 2 arduino where is atm over 60 input used and some outputs.

----- Alkuperäinen viesti -----
Lähettäjä: "Oliver Hall" <notifications@github.com>
Lähetetty: ‎12.‎9.‎2015 10:13
Vastaanottaja: "dcs-bios/dcs-bios" <dcs-bios@noreply.github.com>
Kopio: "Nasty76" <nasty76@gmail.com>
Aihe: Re: [dcs-bios] Dcs-bios + cheap pots (#72)

This pretty much what I wrote in my pull request months ago. I've since moved on and am writing my own integration system.

**Comment by Nasty76** [comment]
Oh well im too noob to code that pot read right , my code work one pot , but after i add more pots , spam start again, i dont know how that share that pot code :(. so i think i need wait if someone write better fix. :D still thanks for this dcs-bios there is so easy to make cockpit inputs in dcs.

**Comment by Nasty76** [comment]
ah ok i think i solve pot issue , even there is multiple pots. i edit file dcsbios.h file and like this:
class Potentiometer : PollingInput {
	private:
		void pollInput();
		char* msg_;
		char pin_;
		unsigned int lastState_;
		int lastsensorValue = 0;
int sensorValue = 0;
int jitter =3; 
	public:
		Potentiometer(char* msg, char pin);
};

and DcsBios.cpp file as this:
Potentiometer::Potentiometer(char* msg, char pin) {

	msg_ = msg;
	pin_ = pin;
	pinMode(pin_, INPUT);
	sensorValue = analogRead(pin_);
	if(sensorValue< lastsensorValue-jitter || sensorValue> lastsensorValue+jitter )
  {
	lastState_ = map(analogRead(pin_), 0, 1023, 0, 65535);
}}
void Potentiometer::pollInput() {
 sensorValue = analogRead(pin_);
  if(sensorValue< lastsensorValue-jitter || sensorValue> lastsensorValue+jitter )
  {
	unsigned int state = map(analogRead(pin_), 0, 1023, 0, 65535);
	if (state != lastState_) {
		char buf[6];
		utoa(state, buf, 10);
		sendDcsBiosMessage(msg_, buf);
}
lastState_ = state;

lastsensorValue = sensorValue;
}
	
}

now i can use multiple pots my radio panels.
